Base/src/main/java/io/deephaven/base/ringbuffer/CharRingBuffer.java

```diff
@@ -0,0 +1,300 @@
+     *
+     * @param capacity minimum capacity of ring buffer
+     */
+    public CharRingBuffer(int capacity) {
+        this(capacity, true);
+    }
+
+    /**
+     * Create a ring buffer of char primitives
+     *
+     * @param capacity minimum capacity of ring buffer
+     * @param growable whether to allow growth when the buffer is full.
+     */
+    public CharRingBuffer(int capacity, boolean growable) {
+        Assert.eqTrue(capacity <= ArrayUtil.MAX_ARRAY_SIZE, "CharRingBuffer size <= MAX_ARRAY_SIZE");
+
+        this.growable = growable;
+        if (growable) {
+            // use next larger power of 2
+            storage = new char[Integer.highestOneBit(capacity - 1) << 1];
+        } else {
+            // might as well use exact size and not over-allocate
+            storage = new char[capacity];
+        }
+
+        tail = head = 0;
+    }
+
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    public int size() {
+        return size;
+    }
+
+    public int capacity() {
+        return storage.length;
+    }
+
+    public int remaining() {
+        return storage.length - size;
+    }
+
+    public void clear() {
+        size = tail = head = 0;
+    }
+
+    /**
+     * Adds an entry to the ring buffer, will throw an exception if buffer is full. For a graceful failure, use
+     * {@link #offer(char)}
+     *
+     * @param e the char to be added to the buffer
+     * @throws UnsupportedOperationException when {@code growable} is {@code false} and buffer is full
+     * @return {@code true} if the char was added successfully
+     */
+    public boolean add(char e) {
+        if (isFull()) {
+            if (!growable) {
+                throw new UnsupportedOperationException("Ring buffer is full and growth is disabled");
+            } else {
+                grow();
+            }
+        }
+        addUnsafe(e);
+        return true;
+    }
+
+    /**
+     * Ensure that there is sufficient empty space to store {@code count} items in the buffer. If the buffer is
+     * {@code growable}, this may result in an internal growth operation. This call should be used in conjunction with
+     * {@link #addUnsafe(char)}.
+     *
+     * @param count the minimum number of empty entries in the buffer after this call
+     * @throws UnsupportedOperationException when {@code growable} is {@code false} and buffer is full
+     */
+    public void ensureRemaining(int count) {
+        if (remaining() < count) {
+            if (!growable) {
+                throw new UnsupportedOperationException("Ring buffer is full and growth is disabled");
+            } else {
+                grow(count);
+            }
+        }
+    }
+
+    /**
+     * Add values unsafely (will silently overwrite values if the buffer is full). This call should be used in
+     * conjunction with {@link #ensureRemaining(int)}.
+     *
+     * @param e the value to add to the buffer
+     */
+    public void addUnsafe(char e) {
```

Comment:
The goal with the unsafe calls is to make a batch of calls as lightweight as possible.  If `isFull()` is that cheap, then I'm happy to include the test but the only current usages of the push/popUnsafe calls follow external guards for ensuring that the buffer will never over or underflow from the subsequent calls.  Also, I thought the mod operation was preferred over an if statement for for branch prediction.  `UpdateBy` needs very fast insertion and removal routines because every value in a processed column will be pushed into and popped out of one of these buffers during creation. Hopefully we can come up with the minimum overhead that keeps us safe.

Reply:
This kind of stuff is often voodoo and oral tradition unfortunately, but I think the received wisdom is that branches that almost-always go the same way are essentially free (thanks to branch prediction). The body of `isFull` is  `return size == storage.length`, and that will surely be inlined; furthermore it is a test that will almost always be false and so it will be essentially free.

I would make the same argument for the mod: divisions and mods are by far the slowest math operations around and again, compared to an 'if' which almost-always evaluates to false, the 'if' is going to win big-time because the CPU is going to speculate right through it and almost always be right.

That's my argument anyway; other wiser minds may want to weigh in.